README.txt

MOVIEDIMMER 1.1


MovieDimmer is a software for dimming the screen around
a user specified area in order to make watching videos in
the dark pleasurable and cozy.


Quickstart step by step:

1. click "Add frame"
2. overlap the video screen area with the measuring window 
3. click "Confirm Selection"
4. name your new frame
5. select the name you specified in the drop down list
6. click activate
7. Enjoy your movie


Observe:

* When the movie dimmers are activated, you can deactivate it
by either clicking on the checkbox next to the "Frame" slider
or by clicking on the frame itself.

* If the "Porthole" slider is deactivated, you can click through
the hole, however if the slider is activated, you cannot click 
through the hole.

* This program has only been tested on Windows 7 with a Java 7
runtime environment. No tests on multiple screens has been done.


Contact:

Name: 		Dennis von Bargen
Email:		dennis@vonbargen.se
Phone:		0735-398898 (Sweden)
Website:	www.vonbargen.se/dennis/

CHANGELOG.txt
-------------------------------------------------
SleepMode 2.0 - 2013-04-22-15:30

SleepMode 1.0 upgraded to SleepMode 2.0!

Removed hardcoded shroudbuttons

Added customizable shrouds
Added load and save functions

Redesigned the GUI


-------------------------------------------------
SleepMode 2.1 - 2013-04-22-15:56

Added sorting to the dropdown list


-------------------------------------------------
SleepMode 2.2 - 2013-04-22-21:04

Fixed default black screen not fitting every screen
Fixed dialogs appearing behind main window


-------------------------------------------------
SleepMode 2.3 - 2013-04-24-18:26

Added transparency to shrouds
Added optional shroud sliders

Redesigned GUI


-------------------------------------------------
MovieDimmers 1.0 (SleepMode 2.4) - 2013-04-25-11:09

Optimized GUI for public release

Added drag and drop features to sliders
Added activate/deactivate checkboxes to sliders
Added minimize button to sliders
Added drag bar to sliders

Renamed SleepMode to MovieDimmers
Renamed shrouds to frames
Renamed window to porthole

-------------------------------------------------
MovieDimmers 1.1 - TBD

Released under a MIT-license

Fixed bugs regarding the default black screen

Added comments and documentation to source code

source/sleepmode/SleepMode.java
package sleepmode;

import java.awt.Color;
import java.awt.Rectangle;
import java.awt.Toolkit;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.ArrayList;
import java.util.Collections;

import javax.swing.JOptionPane;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;

/**
 * SleepMode is a program for creating black boxes around a specified rectangle.
 * It's perfect for creating dimmers for any kind of video site that you are visiting.
 * 
 * @author Dennis von Bargen
 *
 */
public class SleepMode {
	
	static SleepMode s;
	static MainWindow mainWindow;
	static DarkWindow du, dd, dl, dr;	//du = up, dd = down, dl = left, dr = right
	static PeekWindow p;
	static MeasuringWindow measure;
	static TransparencyWindow transOptions;
	
	public static ArrayList<DarkWindowSetting> settings = new ArrayList<DarkWindowSetting>();
	public static String[] settingNames;
	
	private static int selectedSetting;
	
	/**
	 * Initializes the main program
	 */
	public SleepMode() {
		load();
		refresh();
	}
	
	/**
	 * Removes the active dimmers and restores the program to its original state
	 */
	public static void closeAllDark() {
		if (du != null) du.dispose();
		if (dd != null) dd.dispose();
		if (dl != null) dl.dispose();
		if (dr != null) dr.dispose();
		if (p != null) p.dispose();
		if(transOptions != null) transOptions.dispose();
		mainWindow.setVisible(true);
	}
	
	/**
	 * Activates the dimmers
	 * 
	 * @param index			Indicates which frame configuration will be selected
	 * @param sliderOn		Indicates whether the sliders will appear or not
	 * @param optionsOnTop	Indicates where the sliders will appear
	 * 
	 * @return				True if successful, false if there are no frame settings stored
	 */
	public static boolean shroud(int index, boolean sliderOn, boolean optionsOnTop) {
		Object[] shroud;
		selectedSetting = index;
		if (settings.size() > 0) {
			shroud = settings.get(index).getShroud();
		}
		else {
			mainWindow.setAlwaysOnTop(false);
			JOptionPane.showMessageDialog(null, "No frames are stored.");
			mainWindow.setAlwaysOnTop(true);
			return false;
		}
		if (shroud[DarkWindowSetting.UP] != null) {
			du = new DarkWindow((Rectangle)shroud[DarkWindowSetting.UP]);
		}
		if (shroud[DarkWindowSetting.DOWN] != null) {
			dd = new DarkWindow((Rectangle)shroud[DarkWindowSetting.DOWN]);
		}
		if (shroud[DarkWindowSetting.LEFT] != null) {
			dl = new DarkWindow((Rectangle)shroud[DarkWindowSetting.LEFT]);
		}
		if (shroud[DarkWindowSetting.RIGHT] != null) {
			dr = new DarkWindow((Rectangle)shroud[DarkWindowSetting.RIGHT]);
		}
		if(sliderOn) {
			transOptions = new TransparencyWindow();
			transOptions.frameTrans.addChangeListener(s.new FrameTrans());
			transOptions.midTrans.addChangeListener(s.new MidTrans());
			transOptions.frameOn.addActionListener(s.new FrameCheck());
			transOptions.midOn.addActionListener(s.new WindowCheck());
			transOptions.midTrans.setEnabled(false);
			transOptions.mTLabel.setForeground(new Color(0, 0, 155));
			if(optionsOnTop) 
				transOptions.setBounds(dl.getWidth(), (du.getHeight() - transOptions.getHeight()), transOptions.getWidth(), transOptions.getHeight());
			if (du != null) du.setOpacity(75);
			if (dd != null) dd.setOpacity(75);
			if (dl != null) dl.setOpacity(75);
			if (dr != null) dr.setOpacity(75);
			transOptions.frameTrans.setValue(75);
		}
		else {
			du.setOpacity(100);
			dd.setOpacity(100);
			dl.setOpacity(100);
			dr.setOpacity(100);
		}
		return true;
	}
	
	/**
	 * Adds a new shroud to the settings ArrayList and sorts the ArrayList accordingly
	 * 
	 * @param name	The name of the setting
	 * @param port	A rectangle representing the position and size of the porthole
	 * @param up	A rectangle representing the position and size of the upper DarkWindow
	 * @param down	A rectangle representing the position and size of the lower DarkWindow
	 * @param left	A rectangle representing the position and size of the left DarkWindow
	 * @param right	A rectangle representing the position and size of the right DarkWindow
	 */
	public static void addShroud(String name, Rectangle port, Rectangle up, Rectangle down, Rectangle left, Rectangle right) {
		settings.add(new DarkWindowSetting(name, port, up, down, left, right));
		Collections.sort(settings);
		refresh();
	}
	
	/**
	 * Deletes the frame setting that is specified by the index
	 * 
	 * 
	 * @param index		The index of the frame setting to be deleted
	 */
	public static void deleteShroud(int index) {
		int cmd = JOptionPane.showConfirmDialog(null, 
				"Are you sure you want to delete "+settings.get(index).getName()+"?", 
				"Deletion confirmation", 
				JOptionPane.YES_NO_OPTION, 
				JOptionPane.QUESTION_MESSAGE);
		if(cmd == JOptionPane.YES_OPTION) {
			settings.remove(index);
			refresh();
		}
	}
	
	/**
	 * Refreshes the window in order for the list to update
	 */
	public static void refresh() {
		settingNames = null;
		settingNames = new String[settings.size()];
		for(int i = 0; i < settings.size(); i++) {
			settingNames[i] = settings.get(i).getName();
		}
		if (mainWindow != null) {
			mainWindow.dispose();
			mainWindow = null;
			mainWindow = new MainWindow();
		}
	}
	
	@SuppressWarnings("unchecked")
	/**
	 * Loads the settings file.
	 * If it doesn't find the file, a new file is created with the default
	 * frame settings added. After that the method is called recursively
	 * in order to complete the loading
	 */
	public static void load() {
		try{
            FileInputStream fileStream = new FileInputStream("settings.shroud");
            ObjectInputStream in = new ObjectInputStream(fileStream);
            settings = (ArrayList<DarkWindowSetting>)in.readObject();
        }
        catch (FileNotFoundException e) {
            addShroud("Black Screen", 
            		null,
            		new Rectangle(
            				0,
            				0,
            				Toolkit.getDefaultToolkit().getScreenSize().width, 
            				Toolkit.getDefaultToolkit().getScreenSize().height), 
            		null, 
            		null, 
            		null);
            addShroud("Example Frame",
            		new Rectangle(100, 100, 640, 480),
            		new Rectangle(0, 0, Toolkit.getDefaultToolkit().getScreenSize().width, 100),
            		new Rectangle(0, 580, Toolkit.getDefaultToolkit().getScreenSize().width, (Toolkit.getDefaultToolkit().getScreenSize().height - 580)),
            		new Rectangle(0, 100, 100, 480),
            		new Rectangle(740, 100, (Toolkit.getDefaultToolkit().getScreenSize().width - 740), 480));
            save();
            load();
        }
        catch (IOException e) {
            System.out.println("Load IOException");
        }
        catch (ClassNotFoundException e) {
        	System.out.println("Load ClassNotFoundException");
        }
	}
	
	/**
	 * Saves the current settings ArrayList into the settings.shroud file 
	 */
	public static void save() {
		try {
            FileOutputStream fileStream = new FileOutputStream("settings.shroud");
            ObjectOutputStream out = new ObjectOutputStream(fileStream);
            out.writeObject(settings);
        }
        catch (IOException e) {
        	System.out.println("Save IOException");
        }
	}
	
	/**
	 * Checks if the user wishes to have the porthole on.
	 * 
	 * If so, the porthole is created, if not and if the porthole
	 * isn't null, the porthole is disposed
	 */
	public void changeWindowState() {
		boolean windowOn = transOptions.midOn.isSelected();
		if (windowOn) {
			if (settings.get(selectedSetting).getPortHole() != null) {
				p = new PeekWindow(settings.get(selectedSetting).getPortHole());
				p.setOpacity(transOptions.midTrans.getValue());
			}
			transOptions.midTrans.setEnabled(true);
			transOptions.mTLabel.setForeground(Color.BLUE);
		}
		else {
			if(p != null) {
				p.setVisible(false);
				p.dispose();
			}
			transOptions.midTrans.setEnabled(false);
			transOptions.mTLabel.setForeground(new Color(0, 0, 155));
		}
	}
	
	/**
	 * Handles communication between the Frame slider and the actual windows
	 */
	class FrameTrans implements ChangeListener {
		public void stateChanged(ChangeEvent c) {
			if (du != null) du.setOpacity(transOptions.frameTrans.getValue());
			if (dd != null) dd.setOpacity(transOptions.frameTrans.getValue());
			if (dl != null) dl.setOpacity(transOptions.frameTrans.getValue());
			if (dr != null) dr.setOpacity(transOptions.frameTrans.getValue());
		}
	}
	
	/**
	 * Handles communication between the Porthole slider and the actual porthole
	 */
	class MidTrans implements ChangeListener {
		public void stateChanged(ChangeEvent c) {
			if (p != null) p.setOpacity(transOptions.midTrans.getValue());
		}
	}
	
	/**
	 * Checks if the Frame checkbox is selected, if it isn't, the frames will be disposed
	 */
	class FrameCheck implements ActionListener {
		public void actionPerformed(ActionEvent a) {
			if (!transOptions.frameOn.isSelected()) {
				closeAllDark();
			}
		}
	}
	
	/**
	 * Calls changeWindowState if the checkbox is either checked or unchecked
	 */
	class WindowCheck implements ActionListener {
		public void actionPerformed(ActionEvent a) {
			changeWindowState();
		}
	}
	
	public static void main(String[]args) {
		s = new SleepMode();
		mainWindow = new MainWindow();
	}
}


source/sleepmode/DarkWindow.java
package sleepmode;

import java.awt.Color;
import java.awt.Rectangle;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;

import javax.swing.JFrame;
import javax.swing.JPanel;

@SuppressWarnings("serial")
public class DarkWindow extends JFrame {
	
	JPanel black;
	
	public DarkWindow(int x, int y, int width, int height) {
		super();
		setBackground(Color.BLACK);
		addWindowListener(new WindowClose());
		addMouseListener(new ClickClose());
		
		setAlwaysOnTop(true);
		setBounds(x, y, width, height);
		setUndecorated(true);
		setVisible(true);
	}
	
	public DarkWindow(Rectangle rect) {
		super();
		setBackground(Color.BLACK);
		addWindowListener(new WindowClose());
		addMouseListener(new ClickClose());
		
		setAlwaysOnTop(true);
		setBounds(rect);
		setUndecorated(true);
		setVisible(true);
	}
	
	public void setOpacity(int procent) {
		Double opacity = procent * 2.55;
		setBackground(new Color(0, 0, 0, opacity.intValue()));
	}
	
	class WindowClose extends WindowAdapter {
		public void windowClosing(WindowEvent w) {
			SleepMode.closeAllDark();
		}
	}
	
	class ClickClose extends MouseAdapter {
		public void mousePressed(MouseEvent m) {
			SleepMode.closeAllDark();
		}
	}
}


source/sleepmode/MeasuringWindow.java
package sleepmode;

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Font;
import java.awt.event.ComponentAdapter;
import java.awt.event.ComponentEvent;

import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JPanel;

@SuppressWarnings("serial")
public class MeasuringWindow extends JFrame {
	
	JLabel xVal, yVal, wVal, hVal;
	JButton confirm, abort;
	JPanel p;
	
	Font f = new Font("Arial", Font.PLAIN, 20);
	
	public MeasuringWindow() {
		super("Measuring Window");
		setSize(420,320);
		setMinimumSize(new Dimension(420,320));
		setLocationRelativeTo(null);
		
		addComponentListener(new Window());
		
		p = new JPanel();
		p.setLayout(new BorderLayout());
		p.setBackground(Color.BLACK);
		
		p.add(xVal = new JLabel(), BorderLayout.WEST);
		p.add(yVal = new JLabel(), BorderLayout.NORTH);
		p.add(wVal = new JLabel(), BorderLayout.EAST);
		p.add(hVal = new JLabel(), BorderLayout.SOUTH);
		xVal.setFont(f);
		yVal.setFont(f);
		wVal.setFont(f);
		hVal.setFont(f);
		xVal.setForeground(Color.WHITE);
		yVal.setForeground(Color.WHITE);
		wVal.setForeground(Color.WHITE);
		hVal.setForeground(Color.WHITE);
		
		JPanel middle = new JPanel();
		middle.setBackground(Color.BLACK);
		middle.add(confirm = new JButton("Confirm Selection"), BorderLayout.CENTER);
		middle.add(abort = new JButton("Abort Selection"), BorderLayout.CENTER);
		p.add(middle);
		
		add(p);
		update();
		
		setDefaultCloseOperation(DO_NOTHING_ON_CLOSE);
		setVisible(true);
	}
	
	/**
	 * Updates the x,y,width,height lables
	 */
	public void update() {
		xVal.setText("X-Position: "+getBounds().x);
		yVal.setText("Y-Position: "+getBounds().y);
		wVal.setText("Width: "+getBounds().width);
		hVal.setText("Height: "+getBounds().height);
	}
	
	/**
	 * Calls the update method if the window is moved or resized
	 */
	class Window extends ComponentAdapter {
		public void componentMoved(ComponentEvent c) {
			update();
		}
		public void componentResized(ComponentEvent c) {
			validate();
			update();
		}
	}
}


source/sleepmode/DarkWindowSetting.java
package sleepmode;

import java.awt.Rectangle;
import java.io.Serializable;

@SuppressWarnings("serial")
public class DarkWindowSetting implements Comparable<DarkWindowSetting>, Serializable {
	
	private String name;
	private Object[] shroud = new Object[5];
	
	public static final int UP = 0;
	public static final int DOWN = 1;
	public static final int LEFT = 2;
	public static final int RIGHT = 3;
	
	public DarkWindowSetting(String name, Rectangle portHole, Rectangle upBox, Rectangle downBox, Rectangle leftBox, Rectangle rightBox) {
		this.name = name;
		shroud[0] = upBox;
		shroud[1] = downBox;
		shroud[2] = leftBox;
		shroud[3] = rightBox;
		shroud[4] = portHole;
	}
	
	public String getName() {
		return name;
	}
	
	public Rectangle getPortHole() {
		return (Rectangle)shroud[4];
	}
	
	public Object[] getShroud() {
		return shroud;
	}
	
	public int compareTo(DarkWindowSetting d) {
		return name.compareTo(d.getName());
	}
}


source/sleepmode/TransparencyWindow.java
package sleepmode;

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Font;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;

import javax.swing.BorderFactory;
import javax.swing.BoxLayout;
import javax.swing.JButton;
import javax.swing.JCheckBox;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JSlider;

@SuppressWarnings("serial")
public class TransparencyWindow extends JFrame{
	
	private int width;
	private int height;
	
	JLabel fTLabel, mTLabel, sLabel;
	JPanel bar, p1, p2;
	public JButton minimize;
	public JCheckBox frameOn, midOn;
	public JSlider frameTrans, midTrans;
	
	Font font = new Font("Monospaced", Font.PLAIN, 14);
	
	DragNDrop dnd = new DragNDrop();
	
	public TransparencyWindow() {
		super();
		setUndecorated(true);
		setLayout(new BoxLayout(getContentPane(), BoxLayout.Y_AXIS));
		setBackground(null);
		
		bar = new JPanel();
		bar.setLayout(new BorderLayout());
		bar.setBorder(BorderFactory.createLineBorder(new Color(100, 0, 100), 2, false));
		bar.add(sLabel = new JLabel("Sliders"), BorderLayout.CENTER);
		bar.add(minimize = new JButton(" - "), BorderLayout.EAST);
		
		bar.setBackground(Color.BLACK);
		sLabel.setForeground(new Color(100, 0, 100));
		minimize.setBackground(Color.BLACK);
		minimize.setForeground(new Color(100, 0, 100));
		minimize.setFont(new Font("Monospaced", Font.PLAIN, 14));
		minimize.setBorder(BorderFactory.createLineBorder(new Color(100, 0, 100), 2, false));
		
		minimize.addActionListener(new Minimize());
		
		add(bar);
		
		p1 = new JPanel();
		frameTrans = new JSlider();
		frameTrans.setMinimum(1);
		frameTrans.setMaximum(100);
		frameTrans.setValue(100);
		
		p1.add(frameOn = new JCheckBox());
		p1.add(fTLabel = new JLabel("Frame   "));
		p1.add(frameTrans);
		
		add(p1);
		
		p1.setBackground(Color.BLACK);
		fTLabel.setFont(font);
		fTLabel.setForeground(Color.BLUE);
		frameTrans.setBackground(Color.BLACK);
		frameOn.setBackground(Color.BLACK);
		frameOn.setSelected(true);
		
		p2 = new JPanel();
		midTrans = new JSlider();
		midTrans.setMinimum(1);
		midTrans.setMaximum(100);
		midTrans.setValue(1);
		
		p2.add(midOn = new JCheckBox());
		p2.add(mTLabel = new JLabel("Porthole"));
		p2.add(midTrans);
		
		add(p2);
		
		p2.setBackground(Color.BLACK);
		mTLabel.setFont(font);
		mTLabel.setForeground(Color.BLUE);
		midTrans.setBackground(Color.BLACK);
		midOn.setBackground(Color.BLACK);
		
		bar.addMouseListener(dnd);
		bar.addMouseMotionListener(dnd);
		
		pack();
		setAlwaysOnTop(true);
		setVisible(true);
		
		width = getWidth();
		height = getHeight();
	}
	
	/**
	 * Handles the drag and drop functions of the window
	 */
	class DragNDrop extends MouseAdapter {
		
		int xStart;
		int yStart;
		
		public void mousePressed(MouseEvent m) {
			xStart = m.getX();
			yStart = m.getY();
			requestFocusInWindow();
		}
		
		public void mouseDragged(MouseEvent m) {
			int x = getX() - xStart + m.getX();
			int y = getY() - yStart + m.getY();
			setLocation(x,y);
		}
	}
	
	/**
	 * Sets the contents visibile to false, or restores the window
	 * by setting visible to true.
	 */
	class Minimize implements ActionListener {
		public void actionPerformed(ActionEvent a) {
			if (p1.isVisible() && p2.isVisible()) {
				p1.setVisible(false);
				p2.setVisible(false);
				setSize(width, bar.getHeight());
			}
			else {
				p1.setVisible(true);
				p2.setVisible(true);
				setSize(width, height);
			}
		}
	}
}


source/sleepmode/MainWindow.java
package sleepmode;

import java.awt.Color;
import java.awt.Dimension;
import java.awt.Font;
import java.awt.GridLayout;
import java.awt.Rectangle;
import java.awt.Toolkit;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;

import javax.swing.BorderFactory;
import javax.swing.BoxLayout;
import javax.swing.JButton;
import javax.swing.JCheckBox;
import javax.swing.JComboBox;
import javax.swing.JFrame;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.border.Border;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;

@SuppressWarnings("serial")
public class MainWindow extends JFrame {
	
	JCheckBox slideCheck, optionLocation;
	JComboBox<String> setBox;
	JButton activateShroudBtn, addBtn, delBtn;
	MeasuringWindow m;
	
	Border border;
	Font titleFont = new Font("Arial", Font.PLAIN, 14);
	
	public MainWindow() {
		super("MovieDimmers 1.1");
		setLayout(new BoxLayout(getContentPane(), BoxLayout.Y_AXIS));
		
		JPanel sliders = new JPanel();
		sliders.setLayout(new GridLayout(0,2));
		sliders.setBorder(BorderFactory.createTitledBorder(border, "Settings", 0, 0, titleFont, Color.BLUE));
		
		slideCheck = new JCheckBox("Dimmer options sliders");
		slideCheck.addChangeListener(new SliderChecked());
		slideCheck.setSelected(true);
		sliders.add(slideCheck);
		
		optionLocation = new JCheckBox("Sliders over porthole");
		optionLocation.setEnabled(false);
		sliders.add(optionLocation);
		
		add(sliders);
		
		JPanel activate = new JPanel();
		activate.setLayout(new GridLayout(0,1));
		activate.setBorder(BorderFactory.createTitledBorder(border, "Activate!", 0, 0, titleFont, Color.BLUE));
		activateShroudBtn = new JButton("Activate Dimmers");
		activateShroudBtn.addActionListener(new ActivateShroud());
		activate.add(activateShroudBtn);
		
		add(activate);
		
		JPanel settings = new JPanel();
		settings.setLayout(new GridLayout(0,1));
		settings.setBorder(BorderFactory.createTitledBorder(border, "Frames", 0, 0, titleFont, Color.BLUE));
		setBox = new JComboBox<String>(SleepMode.settingNames);
		settings.add(setBox);
		
		addBtn = new JButton("Add Frame");
		addBtn.addActionListener(new AddShroud());
		settings.add(addBtn);
		
		delBtn = new JButton("Delete Frame");
		delBtn.addActionListener(new DeleteShroud());
		settings.add(delBtn);
		
		add(settings);
		
		addWindowListener(new WindowClose());
		
		setSize(400, 300);
		setResizable(false);
		setDefaultCloseOperation(DO_NOTHING_ON_CLOSE);
		setAlwaysOnTop(true);
		setLocationRelativeTo(null);
		setVisible(true);
	}
	
	/**
	 * Enables the optionLocation checkbox if the slideCheck checkbox is checked and vice versa 
	 */
	class SliderChecked implements ChangeListener {
		public void stateChanged(ChangeEvent arg0) {
			if (optionLocation != null) {
				if(slideCheck.isSelected()) {
					optionLocation.setEnabled(true);
				}
				else {
					optionLocation.setEnabled(false);
				}
			}
		}
	}
	
	/**
	 * Activates the shroud and sets this windows visibility to false
	 */
	class ActivateShroud implements ActionListener {
		public void actionPerformed(ActionEvent a) {
			boolean confirm = SleepMode.shroud(setBox.getSelectedIndex(), slideCheck.isSelected(), optionLocation.isSelected());
			if (confirm) {
				setVisible(false);
			}
		}
	}
	
	/**
	 * Sets the program into measuring mode
	 */
	class AddShroud implements ActionListener {
		public void actionPerformed(ActionEvent a) {
			setVisible(false);
			m = new MeasuringWindow();
			m.confirm.addActionListener(new CreateShroud());
			m.abort.addActionListener(new Abort());
		}
	}
	
	/**
	 * Creates a new frame setting based on the measurements made in the
	 * measuring mode and passes the rectangles into the main program
	 */
	class CreateShroud implements ActionListener {
		public void actionPerformed(ActionEvent a) {
			Dimension screen = Toolkit.getDefaultToolkit().getScreenSize();
			String name = "";
			while(name.equals("")) {
				name = JOptionPane.showInputDialog("Name:");
				if(name.equals("")) {
					JOptionPane.showMessageDialog(null, "Please input a name");
				}
			}
			Rectangle port = new Rectangle(m.getX(), m.getY(), m.getWidth(), m.getHeight());
			Rectangle up = new Rectangle(0, 0, screen.width, m.getY());
			Rectangle down = new Rectangle(0, (m.getY() + m.getHeight()), screen.width, (screen.height - (m.getY() + m.getHeight())));
			Rectangle left = new Rectangle(0, m.getY(), m.getX(), m.getHeight());
			Rectangle right = new Rectangle((m.getX() + m.getWidth()), m.getY(), (screen.width - (m.getX() + m.getWidth())), m.getHeight());
			SleepMode.addShroud(name, port, up, down, left, right);
			m.dispose();
		}
	}
	
	/**
	 * Aborts measuring mode
	 */
	class Abort implements ActionListener {
		public void actionPerformed(ActionEvent a) {
			m.dispose();
			setVisible(true);
		}
	}
	
	/**
	 * Activates the main programs delete frame setting method
	 */
	class DeleteShroud implements ActionListener {
		public void actionPerformed(ActionEvent a) {
			setAlwaysOnTop(false);
			SleepMode.deleteShroud(setBox.getSelectedIndex());
			setAlwaysOnTop(true);
		}
	}
	
	/**
	 * Handles the window closing, makes sure that the main program saves
	 * the current settings and exits
	 */
	class WindowClose extends WindowAdapter {
		public void windowClosing(WindowEvent w) {
			SleepMode.save();
			System.exit(0);
		}
	}
}
